docs/userguide/src/migration/prefix.md

```diff
@@ -0,0 +1,187 @@
+        +   It returns `None` if `addr` is not zero.
+        +   If you know `addr` cannot be zero, you can use the new method
+            `from_raw_address_unchecked(addr)`, instead.
+*   module `mmtk::memory_manager`
+    -   **Only affects users of write barriers**
+    -   `object_reference_write_pre(mutator, obj, slot, target)`
+        +   The `target` parameter is now `Option<ObjectReference>`.
+        +   Pass `None` if the slot was holding a NULL reference or any non-reference value.
+    -   `object_reference_write_post(mutator, obj, slot, target)`
+        +   Same as above.
+    -   `object_reference_write(mutator, obj, slot, target)`
+        +   It is labelled as `#[deprecated]` and needs to be redesigned.  It cannot handle the case
+            of storing a non-reference value (such as tagged small integer) into the slot.
+        +   Before a replacement is available, use `object_reference_write_pre` and
+            `object_reference_write_post`, instead.
+
+VM bindings need to re-implement the following traits:
+
+*   trait `Edge`
+    -   `load()`
+        +   The return type is changed to `Option<ObjectReference>`.
+        +   It returns `None` if the slot is holding a NULL reference or other non-reference
+            values.  MMTk will skip those slots.
+*   trait `ObjectModel`
+    -   `copy(from, semantics, copy_context)`
+        +   Previously VM bindings convert the result of `copy_context.alloc_copy()` to
+            `ObjectReference` using `ObjectReference::from_raw_address()`.
+        +   Because `CopyContext::alloc_copy()` never returns zero, you can use
+            `ObjectReference::from_raw_address_unchecked()` to skip the zero check.
+    -   `get_reference_when_copied_to(from, to)`
+        +   `to` is never zero because MMTk only calls this after the destination is determined.
+        +   You may skip the zero check, too.
+    -   `address_to_ref(addr)`
+        +   `addr` is never zero because this method is an inverse operation of
+            `ref_to_address(objref)` where `objref` is never NULL.
+        +   You may skip the zero check, too.
+*   trait `ReferenceGlue`
+    -   *Note: If your VM binding is still using `ReferenceGlue` and the reference processor and
+        finalization processor in mmtk-core, it is strongly recommended to switch to the
+        `Scanning::process_weak_refs` method and implement weak reference and finalization
+        processing on the VM side.*
+    -   `get_referent()`
+        +   The return type is changed to `Option<ObjectReference>`.
+        +   It now returns `None` if the referent is cleared.
+    -   `clear_referent()`
+        +   It now needs to be explicitly implemented because `ObjectReference::NULL` no longer
+            exists.
+        +   Note: The `Edge` trait does not have a method for storing NULL to a slot.  The VM
+            binding needs to implement its own method to store NULL to a slot.
+
+Miscellaneous changes:
```

Comment:
Because strictly speaking, the API is *not* changed w.r.t. `get_finalized_object` and `get_forwarded_object`.  They returned `Option<ObjectReference>` before, and they still return `Option<ObjectReference>`.  However, it is worth mentioning because the binding may expose them to C/C++, and that will fail to compile if the `extern "C"` functions return `ObjectReference` which was previously nullable.

Reply:
I see. You mean the API is compliant but the semantic is changed. Are there any changes other than semantic changes that should be put as 'miscellaneous changes'?

Reply:
The semantics is not changed, either.  `get_finalized_object` used to return `Option<ObjectReference>` which meant "there is no finalized object".  It still returns `Option<ObjectReference>` and still means the same.  The problem is how the VM binding exposes it to C++.  MMTk does not provide an official C/C++ API, so the exposed `extern "C"` wrapper is not controlled by MMTk-core.  I mentioned that because many bindings do that, i.e. they use `ObjectReference::NULL` to encode `None`, which is no longer possible.

Strictly speaking, the root cause of that is that `ObjectReference::NULL` became non-nullable.  The `extern "C"` wrappers of `get_finalized_object` and `get_forwarded_object` are just two possible use cases of `ObjectReference::NULL`.  But since we also introduced a `NullableObjectReference`, we mention that in "miscellaneous changes".

If there are other things that have non-obvious but better way to do, we can mention them in "miscellaneous changes", too.  And of course there are other use cases that I can't foresee now.

Reply:
I changed it to "Not API change, but worth noting".  The purpose should be clear now.